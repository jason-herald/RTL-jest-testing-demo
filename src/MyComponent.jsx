import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import "./MyComponent.css";

function MyComponent({ initialCount, onButtonClick }) {
  const [count, setCount] = useState(initialCount);
  const [error, setError] = useState("");

  useEffect(() => {
    setCount(initialCount ?? 0);
  }, [initialCount]);

  useEffect(() => {
    if (count < 0) {
      setError("Error: Count cannot be negative");
    } else {
      setError("");
    }
  }, [count]);

  return (
    <div>
      {error ? <p className="error-message">{error}</p> : <p>Count: {count}</p>}
      <button
        className="increment-button"
        onClick={() => {
          setCount(count + 1);
          onButtonClick && onButtonClick(count + 1);
        }}
      >
        Increment
      </button>
    </div>
  );
}

MyComponent.propTypes = {
  initialCount: PropTypes.number,
  onButtonClick: PropTypes.func,
};

MyComponent.defaultProps = {
  initialCount: 0,
};

export default MyComponent;
