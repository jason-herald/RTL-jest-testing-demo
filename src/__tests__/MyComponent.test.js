import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import "@testing-library/jest-dom";
expect.extend(toHaveNoViolations);
import MyComponent from "../MyComponent";

// 1. Snapshot Test
it("matches snapshot", () => {
  const { asFragment } = render(<MyComponent />);
  expect(asFragment()).toMatchSnapshot();
});

// 2. Rendering Test
it("renders without errors", () => {
  render(<MyComponent />);
});

// 3. Conditional Rendering Test
it("renders count conditionally", () => {
  const { getByText } = render(<MyComponent initialCount={5} />);
  expect(getByText("Count: 5")).toBeInTheDocument();
});

// 4. Event Handling Test
it("handles click event", () => {
  const { getByText } = render(<MyComponent />);
  fireEvent.click(getByText("Increment"));
  expect(getByText("Count: 1")).toBeInTheDocument();
});

// 5. Callback Function Test
it("calls the onButtonClick callback with the new count value", () => {
  const onButtonClick = jest.fn();
  const { getByText } = render(<MyComponent onButtonClick={onButtonClick} />);
  fireEvent.click(getByText("Increment"));
  expect(onButtonClick).toHaveBeenCalledWith(1);
});

// 6. Lifecycle Method Test 
it("simulates the effect of lifecycle methods", () => {
  const { unmount, getByText, rerender } = render(
    <MyComponent initialCount={5} />
  );
  expect(getByText("Count: 5")).toBeInTheDocument();
  rerender(<MyComponent initialCount={10} />);
  unmount();
});

// 7. Asynchronous Behavior Test 
it("handles asynchronous operations", async () => {
  
});

// 8. Error Handling Test
it("displays errors", () => {
  const { getByText } = render(<MyComponent initialCount={-1} />);
  expect(getByText("Error: Count cannot be negative")).toBeInTheDocument(); 
});

// 9. CSS Styling Test
it("has correct style", () => {
  const { getByText } = render(<MyComponent />);
  const incrementButton = getByText("Increment");
  expect(incrementButton).toHaveClass("increment-button");
});

// 10. Accessibility Test
it("is accessible", async () => {
  const { container } = render(<MyComponent />);
  const results = await axe(container);
  expect(results).toHaveNoViolations();
});

// 11. Props Test
it("renders with different props", async () => {
  const { getByText, rerender, findByText } = render(
    <MyComponent initialCount={5} />
  );
  expect(getByText("Count: 5")).toBeInTheDocument();
  rerender(<MyComponent initialCount={10} />);

 
  const count10Element = await findByText("Count: 10");
  expect(count10Element).toBeInTheDocument();
});



it("updates state correctly", () => {
  const { getByText } = render(<MyComponent initialCount={0} />);
  const incrementButton = getByText("Increment");

 
  expect(getByText("Count: 0")).toBeInTheDocument();

  
  fireEvent.click(incrementButton);
  expect(getByText("Count: 1")).toBeInTheDocument();
});

it("handles edge case of null initial count", () => {
  const { getByText } = render(<MyComponent initialCount={null} />);


  expect(getByText("Count: 0")).toBeInTheDocument();
});
