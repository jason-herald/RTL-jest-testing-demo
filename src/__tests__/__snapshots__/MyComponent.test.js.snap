// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <p>
      Count: 0
    </p>
    <button
      class="increment-button"
    >
      Increment
    </button>
  </div>
</DocumentFragment>
`;
